File: app/src/main/java/com/huiliang/aidemo/util/ApkInstaller.java
package com.huiliang.aidemo.util;

import android.content.ActivityNotFoundException;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Build;
import android.provider.Settings;
import android.util.Log;

import androidx.core.content.FileProvider;

import java.io.File;

/**
 * Time: 2024/7/19
 * Author: muse
 * QQ: 554953278
 * Description:
 * ___       ___       ___       ___
 * /\__\     /\__\     /\  \     /\  \
 * /::L_L_   /:/ _/_   /::\  \   /::\  \
 * /:/L:\__\ /:/_/\__\ /\:\:\__\ /::\:\__\
 * \/_/:/  / \:\/:/  / \:\:\/__/ \:\:\/  /
 * /:/  /   \::/  /   \::/  /   \:\/  /
 * \/__/     \/__/     \/__/     \/__/
 */


public class ApkInstaller {

    public static void installApk(Context context, String filePath) {
        File file = new File(filePath);
        if (!file.exists()) {
            Log.e("ApkInstaller", "APK file does not exist: " + filePath);
            return;
        }

        Log.d("ApkInstaller", "Installing APK: " + filePath);

        Uri apkUri;
        Intent intent = new Intent(Intent.ACTION_INSTALL_PACKAGE);

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            apkUri = FileProvider.getUriForFile(context, context.getPackageName() + ".provider", file);
            intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
        } else {
            apkUri = Uri.fromFile(file);
        }

        intent.setData(apkUri);
        intent.setFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION | Intent.FLAG_ACTIVITY_NEW_TASK);

        try {
            context.startActivity(intent);
        } catch (ActivityNotFoundException e) {
            Log.e("ApkInstaller", "Error starting install activity", e);
            // 如果启动安装失败，可能需要请求安装未知来源应用的权限
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                Intent settingsIntent = new Intent(Settings.ACTION_MANAGE_UNKNOWN_APP_SOURCES);
                settingsIntent.setData(Uri.parse("package:" + context.getPackageName()));
                settingsIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                context.startActivity(settingsIntent);
            }
        }
    }

    public static boolean canRequestPackageInstalls(Context context) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            return context.getPackageManager().canRequestPackageInstalls();
        }
        return true;
    }

    public static void installApkWithProvider(Context context, String filePath) {
        File file = new File(filePath);
        if (file.exists()) {
            Uri apkUri = FileProvider.getUriForFile(context, context.getApplicationContext().getPackageName() + ".fileprovider", file);

            Intent intent = new Intent(Intent.ACTION_INSTALL_PACKAGE);
            intent.setData(apkUri);
            intent.setFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION | Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
            intent.putExtra(Intent.EXTRA_NOT_UNKNOWN_SOURCE, true);
            intent.putExtra(Intent.EXTRA_RETURN_RESULT, true);

            context.startActivity(intent);
        }
    }

}

File: app/src/main/java/com/huiliang/picturesque/util/ApkInstaller.kt
package com.huiliang.picturesque.util

import android.content.ActivityNotFoundException
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Build
import android.provider.Settings
import androidx.core.content.FileProvider
import com.blankj.utilcode.util.LogUtils
import java.io.File

/**
 * Time: 2024/12/30
 * Author: muse
 * QQ: 554953278
 * Description:
 *     ___       ___       ___       ___
 *    /\__\     /\__\     /\  \     /\  \
 *   /::L_L_   /:/ _/_   /::\  \   /::\  \
 *  /:/L:\__\ /:/_/\__\ /\:\:\__\ /::\:\__\
 *  \/_/:/  / \:\/:/  / \:\:\/__/ \:\:\/  /
 *    /:/  /   \::/  /   \::/  /   \:\/  /
 *    \/__/     \/__/     \/__/     \/__/
 */
object ApkInstaller {

    fun installApk(context: Context, filePath: String) {
        val file = File(filePath)
        if (!file.exists()) {
            LogUtils.e("ApkInstaller", "APK file does not exist: $filePath")
            return
        }

        LogUtils.e("ApkInstaller", "Installing APK: $filePath")

        val apkUri: Uri = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            FileProvider.getUriForFile(context, "${context.packageName}.provider", file).also {
                context.grantUriPermission(
                    context.packageName, it, Intent.FLAG_GRANT_READ_URI_PERMISSION
                )
            }
        } else {
            Uri.fromFile(file)
        }

        val intent = Intent(Intent.ACTION_INSTALL_PACKAGE).apply {
            setDataAndType(apkUri, "application/vnd.android.package-archive")
            addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION or Intent.FLAG_ACTIVITY_NEW_TASK)
        }

        try {
            context.startActivity(intent)
        } catch (e: ActivityNotFoundException) {
            LogUtils.e("ApkInstaller", "Error starting install activity", e)
            handleInstallError(context)
        }
    }

    fun canRequestPackageInstalls(context: Context): Boolean {
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            context.packageManager.canRequestPackageInstalls()
        } else {
            true
        }
    }

    fun installApkWithProvider(context: Context, filePath: String) {
        val file = File(filePath)
        if (!file.exists()) {
            LogUtils.e("ApkInstaller", "APK file does not exist: $filePath")
            return
        }

        val apkUri = FileProvider.getUriForFile(
            context, "${context.packageName}.fileprovider", file
        )

        val intent = Intent(Intent.ACTION_INSTALL_PACKAGE).apply {
            setDataAndType(apkUri, "application/vnd.android.package-archive")
            addFlags(
                Intent.FLAG_GRANT_READ_URI_PERMISSION or
                        Intent.FLAG_ACTIVITY_NEW_TASK or
                        Intent.FLAG_ACTIVITY_CLEAR_TASK
            )
            putExtra(Intent.EXTRA_NOT_UNKNOWN_SOURCE, true)
            putExtra(Intent.EXTRA_RETURN_RESULT, true)
        }

        try {
            context.startActivity(intent)
        } catch (e: Exception) {
            LogUtils.e("ApkInstaller", "Error starting install activity", e)
        }
    }

    private fun handleInstallError(context: Context) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val settingsIntent = Intent(Settings.ACTION_MANAGE_UNKNOWN_APP_SOURCES).apply {
                data = Uri.parse("package:${context.packageName}")
                addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            }
            context.startActivity(settingsIntent)
        }
    }
}
